File: archive/boj/9012/yangwj/괄호.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const result = [];

for (let i = 1; i <= Number(input[0]); i++) {
  const str = input[i];
  const stack = [];
  let yes = true;

  for (let char of str) {
    if (char === "(") {
      stack.push(char);
    } else if (char === ")") {
      if (stack.length === 0) {
        yes = false;
        break;
      }
      stack.pop();
    }
  }

  if (stack.length !== 0) {
    yes = false;
  }

  result.push(yes ? "YES" : "NO");
}

console.log(result.join("\n"));

File: archive/boj/9012/ppoddo/괄호.py
import sys
 
def main() -> None:
    input = sys.stdin.readline
    
    T = int(input().strip())
    for _ in range(T):
        ps = input().strip()

        if ps.count('(') != ps.count(')'):
            print("NO")
            continue
        
        # 짝이 보장되었다면 형태 검사
        stack = []
        is_vps = True
        
        for char in ps:
            if char == '(':
                stack.append(char)
            elif char == ')':
                if stack:
                    stack.pop()
                else:
                    is_vps = False
                    break
        
        if stack:
            is_vps = False
        
        print("YES" if is_vps else "NO")
if __name__ == "__main__":
    main()
